### DISCUSSION TITLE
srsRAN gNB with srsUE

### DISCUSSION THREAD

**Discussion by nadiayoza** [srsran/srsRAN_Project#67] [General Help]
Hello, 

I am trying to implement an end-to-end 5G connection using srsRAN Project for the gNB and srsRAN 4G for the UE, as indicated in the tutorial recently released [1]. The UE gets RRC connection to the gNB and it gets an IP address. However, when I try to ping between gNB and UE, the round trip time of the ping takes too long: 

sudo ./srsue ue.conf 
Active RF plugins: libsrsran_rf_uhd.so libsrsran_rf_soapy.so libsrsran_rf_blade.so libsrsran_rf_zmq.so
Inactive RF plugins: 
Reading configuration file ue.conf...

Built in Release mode using commit 655a098b7 on branch master.

Opening 1 channels in RF device=uhd with args=clock=external
Supported RF device list: UHD soapy bladeRF zmq file
[INFO] [UHD] linux; GNU C++ version 9.4.0; Boost_107100; UHD_3.15.0.0-74-ge35f66e8
[INFO] [LOGGING] Fastpath logging disabled at runtime.
Opening USRP channels=1, args: type=b200,master_clock_rate=23.04e6
[INFO] [UHD RF] RF UHD Generic instance constructed
[INFO] [B200] Detected Device: B210
[INFO] [B200] Operating over USB 3.
[INFO] [B200] Initialize CODEC control...
[INFO] [B200] Initialize Radio control...
[INFO] [B200] Performing register loopback test... 
[INFO] [B200] Register loopback test passed
[INFO] [B200] Performing register loopback test... 
[INFO] [B200] Register loopback test passed
[INFO] [B200] Asking for clock rate 23.040000 MHz... 
[INFO] [B200] Actually got clock rate 23.040000 MHz.
Setting manual TX/RX offset to 300 samples
Waiting PHY to initialize ... done!
Attaching UE...
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=5934
Random Access Complete.     c-rnti=0x4601, ta=0
RRC Connected
PDU Session Establishment successful. IP: 10.46.0.36
RRC NR reconfiguration successful.

![image](https://user-images.githubusercontent.com/25062887/235551420-3e4096c0-58d6-4970-ab2a-cee17bc628b2.png)

The amf.yaml is as follows and, based on it, I am using 127.0.0.2 as the AMF address and 127.0.0.100 or 127.0.1.1 as the AMF bind_addr in the gnb.yaml file:

![image](https://user-images.githubusercontent.com/25062887/235552542-18c17f62-48d7-4aa4-bea9-07fa5bbbf824.png)

I tried both ways: conducted tests using physical radios (B210s) connected through a 30 dB attenuator or, instead, using zmq, but I obtain the same results. I am using Ubuntu 20.04 LTS and Open5GS core. I didn't have any problems using the same parameters with the previous srsRAN version. The subscribers parameters indicated in Open5gs are according to the ue.conf parameters. 

[ue.txt](https://github.com/srsran/srsRAN_Project/files/11368793/ue.txt)
[gnb.txt](https://github.com/srsran/srsRAN_Project/files/11368794/gnb.txt)
[upf.txt](https://github.com/srsran/srsRAN_Project/files/11368795/upf.txt)
[amf.txt](https://github.com/srsran/srsRAN_Project/files/11368796/amf.txt)

[1] https://docs.srsran.com/projects/project/en/latest/tutorials/source/srsUE/source/index.html#

If you have any suggestions, please, let me know.

Thanks!

**Comment by robertfalkenberg** [comment]
Hi,
please share the gNB logs, UE logs, and - if possible - PCAP files from the gNB (both, MAC and NGAP).

Best regards

**Reply by nadiayoza** [reply]
Hi Robert, 

Thanks for your response. Please, find below the logs and pcap files:
https://drive.google.com/drive/folders/1aB0OIdCQFcb9r50o4JoEaTxq02DCMr5h?usp=share_link

The ue.log show messages like this:
"2023-05-02T19:30:32.195404 [PHY_LIB] [I] [    0]  PDCCH DMRS Rx: Measure L=2; ncce=4; RSRP=-47.8 dBfs; EPRE=-34.1 dBfs; Corr=0.043
2023-05-02T19:30:32.195406 [PHY_LIB] [I] [    0]  Discarded PDCCH candidate L=2;ncce=4; Correlation is too low (0.0<0.5); EPRE=-34.15; RSRP=-47.83;"

The computers are set to performance mode using sudo ./scripts/srsran_performance
Also, if I just replace the gnb from srsRAN Project with the enb from the new srsRAN 4G 23.04 (the same I'm using for the tests above), the ping round trip time is fine for 5G.

Best regards,

Nadia

**Comment by nadiayoza** [comment]
Hello,

I installed the new srsRAN Project 23.5 released yesterday. Thank you all for the great work! Now the ping round trip time is comparable to the one I got with srsRAN 4G with a maximum throughput of ~23 Mbps. I just wanted to confirm, the adaptive MCS capability is not included yet, right? 

Thanks!

**Reply by frankist** [reply]
We have adaptive MCS capability in the new gnb. It should be active by default. Don't you see the DL and UL mcs change with the channel conditions?

**Reply by nadiayoza** [reply]
Hi Francisco,

Thanks for your response. I was considering the downlink MCS value reported in ue_metrics.csv. Below, I'm attenuating the downlink signal 10 dB every ~10 seconds. The SNR, throughput and BLER change, as expected, but the reported downlink MCS stays at 28. Or is it that the MCS does change, but the value doesn't get updated in the .csv file? In srsRAN with LTE, the MCS value reported would change with the channel conditions, so I was just wondering.

![image](https://github.com/srsran/srsRAN_Project/assets/25062887/baa1f535-deeb-4796-ba4d-5ab54e72801f)

These are conducted tests using B210 for both gNB (with srsRAN Project 23.5) and UE (with srsRAN 4G 23.04). The maximum throughput obtained is 23 Mbps. Is this the expected value?

Thanks

**Reply by frankist** [reply]
Ah weird. Can you share the gnb logs with phy and mac in debug?

**Reply by nadiayoza** [reply]
Hi Francisco, 

Please, find below the log and the ue_metrics.csv file:
https://drive.google.com/file/d/1QUNO_CJTqVmtjMSOPZaKBs_sVdjyKRqh/view?usp=share_link

Now I'm increasing the attenuation 10 dB every 5 seconds (so the log file isn't that large):
![image](https://github.com/srsran/srsRAN_Project/assets/25062887/43fa2e53-0a9b-4cb2-9986-97cc0b3e2898)

Thanks!

**Reply by frankist** [reply]
I took a look at the logs, and it seems that the gnb is always picking mcs=28, because the cqi is always set to 15. E.g.
```
2023-05-10T22:50:09.549962 [UL-PHY2 ] [D] [   671.9] PUCCH: rnti=0x4601 format=2 prb=[2, 3) prb2=na symb=[0, 2) ack=1 sr=1 csi1=1111 t=55.6us
```

`csi1=1111` for this config corresponds to a `cqi=15`. It seems that the limitation is more on the UE than the gNB at this point.

**Reply by nadiayoza** [reply]
Thanks for checking, Francisco! Please, let me know if you guys have any update or suggestion.

**Reply by frankist** [reply]
Hey @nadiayoza, where did you get the `dl_snr` value above? Is this the value reported by the UE?

**Reply by nadiayoza** [reply]
Hi Francisco! Yes, the dl_snr value in the ue_metrics.csv file is reported by the UE. So I was wondering why we see a change in dl_snr and dl_brate, but not in the dl_mcs, which remains in 28. In LTE using srsRAN, the MCS would change as well.
Thanks!

**Reply by frankist** [reply]
Can we retry the same experiment with the latest srsgnb release, and sharing both srsgnb and srsue logs? Having both logs will help me understand if it is not a problem in the PUCCH decoding.

**Reply by nadiayoza** [reply]
Sure, I'll try that and let you know. Thanks!

**Reply by nadiayoza** [reply]
I retried the experiment with the latest versions of srsRAN Project (gNB) and srsRAN 4G (UE). Like in the previous tests, I'm increasing the attenuation 10 dB every 5 seconds. I see a similar behavior, where the MCS doesn't get updated in ue_metrics.csv. Please, find below the gNB and UE logs and the ue_metrics.csv file:
https://drive.google.com/drive/folders/1A853g_bgYMrj5pJCNW9SE1zR1FmfSU4I?usp=sharing
Is it that the MCS is adaptive, but it's just not reported in the ue_metrics.csv? The throughput is decreasing and the BLER is increasing with lower SNR:
![image](https://github.com/srsran/srsRAN_Project/assets/25062887/ff5e4728-51c4-43eb-bd24-fa3615b93412)

Thanks!

**Reply by mochammadfikri28** [reply]
> Lingkaran cahaya,
> 
> Saya memasang srsRAN Project 23.5 baru yang dirilis kemarin. Terima kasih atas kerja kerasnya! Sekarang waktu ping bolak-baliknya sebanding dengan yang saya dapatkan dengan srsRAN 4G dengan throughput maksimum sekitar 23 Mbps. Saya hanya ingin memastikan, kemampuan MCS adaptif belum disertakan, bukan?
> 
> Terima kasih!

Hi @nadiayoza , the project you're working on is similar to the final project I'm currently working on. Could you provide references for the GNB Open5GS and SRS UE files? Thank you.
